File: year2019/day10/sol1.py
import sys
import time
from util import Asteroid, DirectedLine
t1 = time.time()

asteroids = []
with open(sys.argv[1]) as f:
    for y, line in enumerate(f.readlines()):
        for x, el in enumerate(line):
            if el == '#':
                asteroids.append(Asteroid(x, y))        

sol = None
maxi = 0
for a1 in asteroids:
    cnt = 0
    for a2 in asteroids:
        if a1 is a2:
            continue
        line = DirectedLine(a1, a2)
        for other in asteroids:
            if other is a1 or other is a2:
                continue
            
            if other in line:
                break
        
        else:
            cnt += 1
    
    if cnt > maxi:
        sol = a1
        maxi = cnt

print(sol, maxi)
print("time:", time.time() - t1)

File: year2019/day10/java_sol/src/day10/Sol1.java
package day10;

import java.io.FileNotFoundException;
import java.util.List;

import day10.Util.Asteroid;
import day10.Util.DirectedLine;

public class Sol1 {

    public static void main(String[] args) throws FileNotFoundException {
        List<Asteroid> asteroids = Util.getAsteroidsFromFile(args[0]);
        
        int maxi = 0;   
        Asteroid sol = null;
        for (Asteroid a1: asteroids) {
            int cnt = 0;
            for (Asteroid a2: asteroids) {
                if (a1 == a2) {
                    continue;
                }
                DirectedLine line = new DirectedLine(a1, a2);
                boolean visible = true;
                for (Asteroid other: asteroids) {
                    if (other == a1 || other == a2) {
                        continue;
                    }
                    
                    if (line.contains(other)) {
                        visible = false;
                        break;
                    }
                }
                
                if (visible) {
                    ++cnt;
                }
            }
            
            if (cnt > maxi) {
                sol = a1;
                maxi = cnt;
            }
        }

        System.out.println(sol + " " + maxi);
    }
}